### PR TITLE
#29 class check bug fix and #31 path to classes and affinities

### LINKED ISSUE #29: Configuration check -  test availability of classes in affinity.csv
test that all classes in the molecule list are in affinity matrix. The other way around is not necessary as affinity matrix can include all classes. The user selects the classes they want to train for and affinity column of chosen classes is chosen from affinity file

### LINKED ISSUE #31: Receive path of classes.csv and affinity.csv separately
One may need to change the number of classes for various runs and if they are running multiple jobs at the same time using the same dataset, they may need to change their classes.csv file for each run. Therefore the path to this file should be either  received separately or present in the run folder rather than the dataset folder

### PR DESCRIPTION
**Pull request by jolaem**
Fixes #29 
Fixes #31 

Fix bug in previous #29 pull request. Checks for unique ids if no classes.csv is present and in tests.

Add params for affinity and classes to be passed separately. Loss is aware of it and will return Runtime Error if gamma set but affinity matrix not passed.

### CODE REVIEW

File: avae/data.py [hunk 1]
```diff
@@ -151,39 +126,45 @@ class ProteinDataset(Dataset):
...
+                        np.asarray(ids)[~class_check]
+                    )
+                )
 
-            self.paths = [p for p in self.paths for c in class_list if c in p]
+        if classes is not None:
+            with open(classes, newline="\n") as f:
+                classes = np.asarray(f.read().splitlines()).flatten()
+                classes = [c for c in classes if c != ""]
+                classes = [c.strip() for c in classes if len(c.strip()) != 0]
```
Comment (crangelsmith) [inline avae/data.py:162]: In a chat with @marjanfamili yesterday we thought it could be useful to have the list of final classes as an attribute to the ProteinDataset, so it can be accessed later (in [here](https://github.com/alan-turing-institute/affinity-vae/blob/705f4850990165f5bb7faa6bc4db75be1d4db89d/avae/data.py#L40) to visualise the matrix for the given classes). It could be defined after this block of code as `self.final_classes = classes` if there is a classes.csv file, or `self.final_classes = ids`. (doesn't have to be called final_classes btw, is just to avoid confusion with the other variables).

File: run.py [hunk 2]
```diff
@@ -33,6 +33,20 @@
     help="Do not drop last validate batch if "
     "if it is smaller than batch_size.",
 )
+@click.option(
+    "--affinity",
+    "-af",
+    type=str,
+    default=None,
+    help="Path to affinity matrix for training.",
```
Comment (crangelsmith) [inline run.py:41]: Same as comment above, need to be consistent if is a path or filename.

File: run.py [hunk 3]
```diff
@@ -33,6 +33,20 @@
...
+)
+@click.option(
+    "--classes",
+    "-cl",
+    type=str,
+    default=None,
+    help="Path to a CSV file containing a list of classes for training.",
```
Comment (crangelsmith) [inline run.py:48]: Same as the comment above, need to be consistent if is a path or filename.

File: avae/loss.py [hunk 4]
```diff
@@ -97,6 +97,12 @@ def __init__(self, device, beta, gamma=1, lookup_aff=None, recon_fn="MSE"):
         if lookup_aff is not None:
             self.gamma = gamma
             self.affinity_loss = AffinityLoss(lookup_aff, device)
+        elif lookup_aff is None and gamma != 0:
```
Comment (crangelsmith) [inline avae/loss.py:100]: this means that if we don't provide an affinity.csv file, the code has to be setup to be a standard vae, right? Might be useful to to document that somewhere outside the code.

File: avae/data.py [hunk 5]
```diff
@@ -55,14 +18,23 @@ def load_data(
...
     eval: bool = True,
+    affinity=None,
+    classes=None,
 ):
     if not eval:
-        # load affinity matrix
-        lookup = load_affinities(datapath)
+        if affinity is not None:
+            # load affinity matrix
+            lookup = pd.read_csv(os.path.join(datapath, affinity), header=0)
```
[the remark this answers is not in the document]
Comment (jolaem) [reply avae/data.py:27]: Ah, good catch, this is leftover after load_affinity function that was no longer needed. Removing datapath for consistency with classes parameter.

File: avae/data.py (hunk 1 above)
Comment (jolaem) [reply avae/data.py:162]: That sounds good, adding self.final_classes attribute.

File: avae/loss.py (hunk 4 above)
Comment (jolaem) [reply avae/loss.py:100]: Yep, that's correct. If you set --gamma 0 and -pd 0 that runs just a standard beta-VAE. This used to be quite redundant and  if the matrix was present it used to actually calculate the affinity loss and multiply it with gamma=0 so it wouldn't contribute to the total loss but this change pretty much removes this reduncancy.

I'll add a note of this to the README.md.

File: run.py (hunk 2 above)
Comment (jolaem) [reply run.py:41]: Sounds good, I'll patch to remove the legacy os.path.join with datapath and so this will actually use a full path for consistency with classes.

File: run.py (hunk 3 above)
Comment (jolaem) [reply run.py:48]: In this case it actually refers to the full path on [line 159](https://github.com/alan-turing-institute/affinity-vae/blob/705f4850990165f5bb7faa6bc4db75be1d4db89d/avae/data.py#L159). I'll leave it at that since it is now consistent with affinities with the upcoming patch.

**Review by crangelsmith — APPROVED** [review]
